src/records.jl

```diff
@@ -26,49 +28,91 @@ Simply wraps the value `x` in a `Nullable` and sticks that in an `Attribute` wit
 empty `Function`.
 """
 Attribute(x::T) where {T} = Attribute{T}(() -> x)
+Attribute{T}(x::T) where {T} = Attribute{T}(() -> x)
 
 """
     get(attr::Attribute{T}) -> T
 
 Run set `attr.x` to the output of `attr.f` if `attr.x` is not already set.
 We then return the value stored in `attr.x`
 """
-function Base.get(attr::Attribute)
+function Base.get(attr::Attribute{T}) where T
     if isnull(attr.x)
-        attr.x = Nullable(attr.f())
+        attr.x = Nullable{T}(attr.f())
     end
 
-    return get(attr.x)
+    return get(attr.x)::T
 end
 
+hasfield(T::DataType, name::Symbol) = Base.fieldindex(T, name, false) > 0
+
 """
     Record
 
-Are an `Attribute` container used to store information about a log events including the
-msg, date, level, stacktrace, etc. `Formatter`s use `Records` to format log
+A dictionary-like container with `Symbol` keys used to store information about a log events
+including the msg, date, level, stacktrace, etc. `Formatter`s use `Records` to format log
 message strings.
 
-NOTE: you should access `Attribute`s in a `Record` by using `getindex` (ie: record[:msg])
-as this will correctly extract the value from the `Attribute` container.
+You can access the properties of a `Record` by using `getindex` (ie: record[:msg]).
+
+Subtypes of `Record` should implement `getindex(::MyRecord, ::Symbol)` and key-value pair
+iteration.
+"""
+abstract type Record <: AbstractDict{Symbol, Any} end
+
 """
-abstract type Record end
+    AttributeRecord <: Record
+
+A `Record` which stores its properties as `Attribute`s for lazy evaluation.
+
+Calling `getindex` or iterating will evaluate and cache the properties accessed.
+
+Subtypes of `AttributeRecord` should implement `Memento.getattribute(::MyRecord, ::Symbol)`
+instead of `getindex`.
+"""
+abstract type AttributeRecord <: Record end
+
+level(rec::Record) = rec[:level]
+Base.getindex(rec::Record, attr::Symbol) = getfield(rec, attr)
+Base.haskey(rec::Record, attr::Symbol) = hasfield(rec, attr)
+Base.keys(rec::T) where {T <: Record} = (fieldname(T, i) for i in 1:nfields(rec))
+
+Base.start(rec::Record) = 0
+Base.done(rec::T, state) where {T <: Record} = state >= nfields(T)
+
+function Base.next(rec::T, state::Int) where T <: Record
+    new_state = state + 1
+    return (fieldname(T, new_state) => getfield(rec, new_state), new_state)
+end
+
+function Base.next(rec::T, state::Int) where T <: AttributeRecord
+    new_state = state + 1
+    return (fieldname(T, new_state) => get(getfield(rec, new_state)), new_state)
+end
+
+"""
+    getattribute(rec::AttributeRecord, attr::Symbol)
+"""
+getattribute(rec::AttributeRecord, attr::Symbol) = getfield(rec, attr)
```

Comment:
Seems like we could just use `getfield` for this? Alternatively, we can define this as `Base.getproperty` (with an appropriate version check) to allow `rec.attr` access.

Reply:
...except that `a.b` falls back to `getfield` with no `getproperty` method, so I think we can just use `rec.attr` in any case.

Reply:
We want people to be able to override this for their records. E.g., a record which looks things up in multiple places. I think `getproperty` is a good solution but I also think we should wait until Memento drops 0.6 to make that change rather than having a weird multiversion story for custom record types.

Reply:
Yeah, good call. Are you thinking that once we require 0.7 at a minimum we should deprecate this in favor of `getproperty`, or keep them both?

Reply:
We could do that but I'm not sure. We'll see what Rory says.

Reply:
Yeah, I'd be in favour of using `getproperty` once Memento is 0.7 only. I'm still not a fan of `getproperty` overloading, but it seems like that'll be a common julia convention going forward.

Reply:
https://github.com/invenia/Memento.jl/issues/87

Reply:
Agreed regarding the unfortunate introduction of `.` overloading, but I'll note that it is currently possible to use `rec.attr` given that it's simply a call to `getfield`, which is how this function is currently defined anyway.